### PR TITLE
Upgrade to 0.12

### PR DESCRIPTION
**Pull request by maximmi**
## what
* Port (rewrite) module with 0.12 syntax
* Update Codefresh badge to point to the test pipeline in `terraform-modules` project
* Update examples
* Update README 
* Update README examples

## why
* Module currently does not work with 0.12
* Much easier syntax
* `terraform-modules` project contains pipelines for all terraform modules
* closes #18 
* closes #16 

## references
* https://github.com/gruntwork-io/terratest/pull/370

### CODE REVIEW

File: examples/complete/fixtures.us-west-1.tfvars [hunk 1]
```diff
@@ -0,0 +1,9 @@
+namespace = "eg"
```
Comment (aknysh) [inline examples/complete/fixtures.us-west-1.tfvars]: use `us-east-2` for all tests
https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment/tree/master/examples/complete

`us-west-1` is very limited and has different AZs names for diff AWS accounts

Also update the region in Go test

File: examples/complete/fixtures.us-west-1.tfvars [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+stage = "test"
+
+name = "multi-az-subnets-only-private"
+
+availability_zones = ["us-west-2a", "us-west-2b", "us-west-2c"]
```
Comment (aknysh) [inline examples/complete/fixtures.us-west-1.tfvars]:
```suggestion
availability_zones = ["us-east-2a", "us-east-2b", "us-east-2c"]
```

File: examples/complete/main.tf [hunk 3]
```diff
@@ -0,0 +1,40 @@
+locals {
+  public_cidr_block  = cidrsubnet(var.cidr_block, 1, 0)
+  private_cidr_block = cidrsubnet(var.cidr_block, 1, 1)
+}
+
+module "vpc" {
+  source     = "git::https://github.com/cloudposse/terraform-aws-vpc.git?ref=tags/0.7.0"
```
Comment (aknysh) [inline examples/complete/main.tf]:
```suggestion
  source     = "git::https://github.com/cloudposse/terraform-aws-vpc.git?ref=tags/0.8.0"
```

File: examples/complete/main.tf [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  source              = "../../"
+  namespace           = var.namespace
+  stage               = var.stage
+  name                = var.name
+  availability_zones  = var.availability_zones
+  vpc_id              = module.vpc.vpc_id
+  cidr_block          = local.public_cidr_block
+  type                = "public"
+  igw_id              = module.vpc.igw_id
+  nat_gateway_enabled = "true"
```
Comment (aknysh) [inline examples/complete/main.tf]:
```suggestion
  nat_gateway_enabled = true
```

Use TF 0.12 first class types eveywhere

File: private.tf [hunk 5]
```diff
@@ -1,74 +1,107 @@
...
 
 resource "aws_network_acl" "private" {
-  count      = "${var.enabled == "true" && var.type == "private" && signum(length(var.private_network_acl_id)) == 0 ? 1 : 0}"
-  vpc_id     = "${var.vpc_id}"
-  subnet_ids = ["${aws_subnet.private.*.id}"]
-  egress     = "${var.private_network_acl_egress}"
-  ingress    = "${var.private_network_acl_ingress}"
-  tags       = "${module.private_label.tags}"
-  depends_on = ["aws_subnet.private"]
+  count      = var.enabled == "true" && var.type == "private" && signum(length(var.private_network_acl_id)) == 0 ? 1 : 0
```
Comment (aknysh) [inline private.tf]:
```suggestion
  count      = var.enabled == true && var.type == "private" && signum(length(var.private_network_acl_id)) == 0 ? 1 : 0
```

File: public.tf [hunk 6]
```diff
@@ -1,80 +1,115 @@
 locals {
-  public_count              = "${var.enabled == "true" && var.type == "public" ? length(var.availability_zones) : 0}"
-  public_nat_gateways_count = "${var.enabled == "true" && var.type == "public" && var.nat_gateway_enabled == "true" ? length(var.availability_zones) : 0}"
+  public_count              = var.enabled == "true" && var.type == "public" ? length(var.availability_zones) : 0
```
Comment (aknysh) [inline public.tf]:
```suggestion
  public_count              = var.enabled && var.type == "public" ? length(var.availability_zones) : 0
```

File: public.tf [hunk 7]
```diff
@@ -1,80 +1,115 @@
...
 
 resource "aws_network_acl" "public" {
-  count      = "${var.enabled == "true" && var.type == "public" && signum(length(var.public_network_acl_id)) == 0 ? 1 : 0}"
-  vpc_id     = "${var.vpc_id}"
-  subnet_ids = ["${aws_subnet.public.*.id}"]
-  egress     = "${var.public_network_acl_egress}"
-  ingress    = "${var.public_network_acl_ingress}"
-  tags       = "${module.public_label.tags}"
-  depends_on = ["aws_subnet.public"]
+  count      = var.enabled == "true" && var.type == "public" && signum(length(var.public_network_acl_id)) == 0 ? 1 : 0
```
Comment (aknysh) [inline public.tf]:
```suggestion
  count      = var.enabled && var.type == "public" && var.public_network_acl_id == "" ? 1 : 0
```

File: public.tf [hunk 8]
```diff
@@ -1,80 +1,115 @@
...
+  vpc_id     = var.vpc_id
+  subnet_ids = aws_subnet.public.*.id
+  dynamic "egress" {
+    for_each = var.public_network_acl_egress
+    content {
+      # TF-UPGRADE-TODO: The automatic upgrade tool can't predict
```
Comment (aknysh) [inline public.tf]: remove the comments generated by TF

File: public.tf [hunk 9]
```diff
@@ -1,80 +1,115 @@
...
+      ipv6_cidr_block = lookup(egress.value, "ipv6_cidr_block", null)
+      protocol        = lookup(egress.value, "protocol", null)
+      rule_no         = lookup(egress.value, "rule_no", null)
+      to_port         = lookup(egress.value, "to_port", null)
+    }
+  }
+  dynamic "ingress" {
+    for_each = var.public_network_acl_ingress
+    content {
+      # TF-UPGRADE-TODO: The automatic upgrade tool can't predict
```
Comment (aknysh) [inline public.tf]: remove the comments

File: test/src/examples_complete_test.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+// Test the Terraform module in examples/complete using Terratest.
+func TestExamplesComplete(t *testing.T) {
+	t.Parallel()
+
+	terraformOptions := &terraform.Options{
+		// The path to where our Terraform code is located
+		TerraformDir: "../../examples/complete",
+		Upgrade:      true,
+		// Variables to pass to our Terraform code using -var-file options
+		VarFiles: []string{"fixtures.us-west-1.tfvars"},
```
Comment (aknysh) [inline test/src/examples_complete_test.go]:
```suggestion
		VarFiles: []string{"fixtures.us-east-2.tfvars"},
```

File: test/src/examples_complete_test.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+	// Run `terraform output` to get the value of an output variable
+	privateRouteTableIds := terraform.OutputMap(t, terraformOptions, "private_az_route_table_ids")
+	// Run `terraform output` to get the value of an output variable
+	publicNATGateWayIds := terraform.OutputMap(t, terraformOptions, "public_az_ngw_ids")
+	// Run `terraform output` to get the value of an output variable
+	publicRouteTableIds := terraform.OutputMap(t, terraformOptions, "public_az_route_table_ids")
+	// Run `terraform output` to get the value of an output variable
+	publicSubnetIds := terraform.OutputMap(t, terraformOptions, "public_az_subnet_ids")
+
+  expectedAZs := []string {"us-west-2a", "us-west-2b", "us-west-2c"}
```
Comment (aknysh) [inline test/src/examples_complete_test.go]:
```suggestion
  expectedAZs := []string {"us-east-2a", "us-east-2b", "us-east-2c"}
```

File: test/src/examples_complete_test.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	// Verify we're getting back the outputs we expect
+  assert.Equal(t, expectedAZs, getKeys(privateSubnetIds))
```
Comment (aknysh) [inline test/src/examples_complete_test.go]: why all tests use `expectedAZs`?

File: test/src/examples_complete_test.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  assert.Equal(t, expectedAZs, getKeys(privateRouteTableIds))
+  assert.Equal(t, expectedAZs, getKeys(publicNATGateWayIds))
+  assert.Equal(t, expectedAZs, getKeys(publicRouteTableIds))
+  assert.Equal(t, expectedAZs, getKeys(publicSubnetIds))
+}
```
Comment (aknysh) [inline test/src/examples_complete_test.go]: format Go code

File: README.md [hunk 14]
```diff
@@ -52,51 +52,46 @@ It's 100% Open Source and licensed under the [APACHE2](LICENSE).
...
 
 module "public_subnets" {
   source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
-  namespace           = "${var.namespace}"
-  stage               = "${var.stage}"
-  name                = "${var.name}"
+  namespace           = var.namespace
+  stage               = var.stage
+  name                = var.name
   availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
```
Comment (aknysh) [inline README.md]:
```suggestion
  availability_zones  = ["us-east-2a", "us-east-2b", "us-east-2c"]
```

File: README.md [hunk 15]
```diff
@@ -52,51 +52,46 @@ It's 100% Open Source and licensed under the [APACHE2](LICENSE).
...
-  # Need to explicitly provide the count since Terraform currently can't use dynamic count on computed resources from different modules
-  # https://github.com/hashicorp/terraform/issues/10857
-  # https://github.com/hashicorp/terraform/issues/12125
-  # https://github.com/hashicorp/terraform/issues/4149
-  az_ngw_count = 3
+  source             = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
+  namespace          = var.namespace
+  stage              = var.stage
+  name               = var.name
+  availability_zones = ["us-east-1a", "us-east-1b", "us-east-1c"]
```
Comment (aknysh) [inline README.md]:
```suggestion
  availability_zones = ["us-east-2a", "us-east-2b", "us-east-2c"]
```

File: README.md [hunk 16]
```diff
@@ -110,49 +105,48 @@ Given the following configuration
...
 
 module "public_subnets" {
   source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
-  namespace           = "${var.namespace}"
-  stage               = "${var.stage}"
-  name                = "${var.name}"
+  namespace           = var.namespace
+  stage               = var.stage
+  name                = var.name
   availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
```
Comment (aknysh) [inline README.md]: update availability_zones

File: README.md [hunk 17]
```diff
@@ -110,49 +105,48 @@ Given the following configuration
...
 
 module "private_subnets" {
   source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
-  namespace           = "${var.namespace}"
-  stage               = "${var.stage}"
-  name                = "${var.name}"
+  namespace           = var.namespace
+  stage               = var.stage
+  name                = var.name
   availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
```
Comment (aknysh) [inline README.md]: update availability_zones

File: README.yaml [hunk 18]
```diff
@@ -77,50 +77,43 @@ description: |-
...
 
   module "public_subnets" {
     source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
-    namespace           = "${var.namespace}"
-    stage               = "${var.stage}"
-    name                = "${var.name}"
+    namespace           = var.namespace
+    stage               = var.stage
+    name                = var.name
     availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
```
Comment (aknysh) [inline README.yaml]: update availability_zones

File: README.yaml [hunk 19]
```diff
@@ -77,50 +77,43 @@ description: |-
...
-    # Need to explicitly provide the count since Terraform currently can't use dynamic count on computed resources from different modules
-    # https://github.com/hashicorp/terraform/issues/10857
-    # https://github.com/hashicorp/terraform/issues/12125
-    # https://github.com/hashicorp/terraform/issues/4149
-    az_ngw_count = 3
+    source             = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
+    namespace          = var.namespace
+    stage              = var.stage
+    name               = var.name
+    availability_zones = ["us-east-1a", "us-east-1b", "us-east-1c"]
```
Comment (aknysh) [inline README.yaml]: update availability_zones

File: examples/complete/variables.tf [hunk 20]
```diff
@@ -1,25 +1,26 @@
...
 variable "cidr_block" {
-  type        = "string"
+  type        = string
   default     = "10.0.0.0/16"
   description = "Base CIDR block which is divided into subnet CIDR blocks (e.g. `10.0.0.0/16`)"
 }
+
+variable "availability_zones" {
+  type        = list(string)
+  default     = []
```
Comment (aknysh) [inline examples/complete/variables.tf]: remove default, `make availability_zones` required

File: variables.tf [hunk 21]
```diff
@@ -1,38 +1,38 @@
...
-  type        = "map"
+  type        = map(string)
   default     = {}
   description = "Additional tags (e.g. map(`BusinessUnit`,`XYZ`)"
 }
 
 variable "availability_zones" {
-  type        = "list"
+  type        = list(string)
   default     = []
```
Comment (aknysh) [inline variables.tf]: remove default, `make availability_zones` required

File: README.md (hunk 15 above)
Comment (osterman) [reply README.md]: Also, move to fixtures the way we do in all the other 0.12 modules

File: test/src/examples_complete_test.go (hunk 12 above)
Comment (maximmi) [reply test/src/examples_complete_test.go]: @aknysh  this is how the outputs looks like:
```
   Outputs:

   private_az_route_table_ids = {
     "us-east-2a" = "rtb-0489137a5c668e49b"
     "us-east-2b" = "rtb-083c0e942abb4b8a1"
     "us-east-2c" = "rtb-0c36484693db5e774"
   }
   private_az_subnet_ids = {
     "us-east-2a" = "subnet-0f56deccfe81c0ea0"
     "us-east-2b" = "subnet-05861d30d45e7b675"
     "us-east-2c" = "subnet-036d747a2b46857ae"
   }
   public_az_ngw_ids = {
     "us-east-2a" = "nat-0f5057f09b8cd8ddc"
     "us-east-2b" = "nat-0971b2505ea6d03f1"
     "us-east-2c" = "nat-0dc1cdf91010be057"
   }
   public_az_route_table_ids = {
     "us-east-2a" = "rtb-0642afb4401f1eef1"
     "us-east-2b" = "rtb-04f511a28a2d5a6a2"
     "us-east-2c" = "rtb-05f0ee4e831b05697"
   }
   public_az_subnet_ids = {
     "us-east-2a" = "subnet-0dcb9e32f1f02a367"
     "us-east-2b" = "subnet-0b432a6748ca40638"
     "us-east-2c" = "subnet-00a9a6636ca722474"
   }
```
so checking for `keys` to be proper for all resources

**Pull request merged**